src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs

```diff
@@ -660,5 +695,155 @@ public void GetResult()
+            }
+
+            /// <summary>
+            /// Gets a value indicating whether the antedent <see cref="Task"/> has already completed.
+            /// </summary>
+            public bool IsCompleted => this.antecedent.IsCompleted;
+
+            /// <summary>
+            /// Rethrows any exception thrown by the antecedent.
+            /// </summary>
+            public void GetResult() => this.antecedent.GetAwaiter().GetResult();
+
+            /// <summary>
+            /// Schedules a callback to run when the antecedent task completes.
+            /// </summary>
+            /// <param name="continuation">The callback to invoke.</param>
+            public void OnCompleted(Action continuation)
+            {
+                Requires.NotNull(continuation, nameof(continuation));
+
+                this.antecedent.ContinueWith(
+                    (_, s) => ((Action)s)(),
+                    continuation,
+                    CancellationToken.None,
+                    TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default);
+            }
+        }
+
+        /// <summary>
+        /// A Task awaitable that has affinity to executing callbacks synchronously on the completing callstack.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the awaited <see cref="Task"/>.</typeparam>
+        public struct ExecuteContinuationSynchronouslyAwaitable<T>
+        {
+            /// <summary>
+            /// The task whose completion will execute the continuation.
+            /// </summary>
+            private readonly Task<T> antecedent;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ExecuteContinuationSynchronouslyAwaitable{T}"/> struct.
+            /// </summary>
+            /// <param name="antecedent">The task whose completion will execute the continuation.</param>
+            public ExecuteContinuationSynchronouslyAwaitable(Task<T> antecedent)
+            {
+                Requires.NotNull(antecedent, nameof(antecedent));
+                this.antecedent = antecedent;
+            }
+
+            /// <summary>
+            /// Gets the awaiter.
+            /// </summary>
+            /// <returns>The awaiter</returns>
+            public ExecuteContinuationSynchronouslyAwaiter<T> GetAwaiter() => new ExecuteContinuationSynchronouslyAwaiter<T>(this.antecedent);
+        }
+
+        /// <summary>
+        /// A Task awaiter that has affinity to executing callbacks synchronously on the completing callstack.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the awaited <see cref="Task"/>.</typeparam>
+        public struct ExecuteContinuationSynchronouslyAwaiter<T> : INotifyCompletion
+        {
+            /// <summary>
+            /// The task whose completion will execute the continuation.
+            /// </summary>
+            private readonly Task<T> antecedent;
```

Comment:
❔ Why hold the `Task<T>` here instead of just the awaiter?

Reply:
I could. I don't see the difference though. It couldn't possibly be *smaller* than holding a `Task`, but it could theoretically be bigger. I might research that, but it would seem I have nothing to gain by switching it. Are you aware of something that I'm missing?

Reply:
Actually, I couldn't just use the awaiter. The key point of these awaiters is to override the behavior of `TaskAwaiter`. I can't do that if I'm tunneling through `TaskAwaiter`. I have to actually schedule a task continuation explicitly in order to set `TaskContinuationOptions.ExecuteSynchronously`, which I cannot do via `TaskAwaiter.OnCompleted`.